src/ts/datasources/choropleth.ts

```diff
@@ -3,44 +3,115 @@ module powerbi.extensibility.visual.data {
     export class Choropleth extends Datasource {
         private choroplethData: any[];
         private fillColorLimits: mapboxUtils.Limits;
+        private fillSizeLimits: mapboxUtils.Limits;
+        private bboxCache: BBoxCache;
+
+        private static readonly BBOX_TIMEOUT = 1500
+        private static readonly BBOX_TIMER = 'choropleth-bbox-timer'
 
         constructor() {
-            super('choropleth');
+            super('choropleth-source');
+            this.bboxCache = new BBoxCache()
         }
 
-        addSources(map, settings) {
-            map.addSource('choropleth-source', {
+        addSources(map, settings: MapboxSettings) {
+            map.addSource(this.ID, {
                 type: 'vector',
-                url: settings.choropleth[`vectorTileUrl${settings.choropleth.currentLevel}`],
+                url: settings.choropleth.getCurrentVectorTileUrl(),
             });
-            return map.getSource('choropleth-source');
+            return map.getSource(this.ID);
         }
 
         removeSources(map) {
-            map.removeSource('choropleth-source');
+            map.removeSource(this.ID);
         }
 
         ensure(map, layerId, settings): void {
             super.ensure(map, layerId, settings)
-            const source: any = map.getSource('choropleth-source');
+            const source: any = map.getSource(this.ID);
             if (!source) {
                 this.addToMap(map, settings);
             }
         }
 
         getLimits() {
-            return this.fillColorLimits;
+            return {
+                color: this.fillColorLimits,
+                size: this.fillSizeLimits
+            }
         }
 
-        getData(map, settings) : any[] {
+        getData(map, settings): any[] {
             return this.choroplethData;
         }
 
-        update(map, features, roleMap, settings) {
+        update(map, features, roleMap, settings: MapboxSettings) {
             super.update(map, features, roleMap, settings)
+            this.choroplethData = features.map(f => f.properties)
+            this.fillColorLimits = mapboxUtils.getLimits(this.choroplethData, roleMap.color ? roleMap.color.displayName : '')
+            this.fillSizeLimits = mapboxUtils.getLimits(this.choroplethData, roleMap.size ? roleMap.size.displayName : '')
+            const featureNames = this.choroplethData.map(f => f[roleMap.location.displayName])
+            const apiSettings = settings.api
+
+            // NOTE: this is a workaround because 'sourcedata' event of mapbox is received multiple times
```

Comment:
We should revisit the need for this logic in the near future.  We want our Mapboxgl map event model to support this use case without the need for a timer, which will potentially introduce bugs in the future.

Reply:
No necessary to change to merge this PR